File: backend/src/test/java/ca/metricalsky/winston/entity/fetch/FetchRequestEntityTest.java
package ca.metricalsky.winston.entity.fetch;

import ca.metricalsky.winston.entity.fetch.FetchOperationEntity.Type;
import ca.metricalsky.winston.entity.fetch.FetchRequestEntity.Status;
import ca.metricalsky.winston.test.TestUtils;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

@DataJpaTest
class FetchRequestEntityTest {

    @Autowired
    private TestEntityManager entityManager;

    @Test
    void persists() {
        var fetchRequestEntity = FetchRequestEntity.builder()
                .build();

        var persistedEntity = entityManager.persistFlushFind(fetchRequestEntity);

        assertThat(persistedEntity)
                .hasFieldOrPropertyWithValue("status", Status.ACCEPTED)
                .hasNoNullFieldsOrProperties();
        assertThat(persistedEntity.getOperations())
                .isEmpty();
    }

    @Test
    void persistsOperations() {
        var fetchOperationEntity = FetchOperationEntity.builder()
                .operationType(Type.CHANNELS)
                .objectId(TestUtils.randomId())
                .build();
        var fetchRequestEntity = FetchRequestEntity.builder()
                .operations(List.of(fetchOperationEntity))
                .build();

        var persistedEntity = entityManager.persistFlushFind(fetchRequestEntity);

        assertThat(persistedEntity.getOperations())
                .hasSize(1);

        var persistedOperation = entityManager.refresh(persistedEntity.getOperations().getFirst());

        assertThat(persistedOperation.getId())
                .isNotNull();
        assertThat(persistedOperation.getFetchRequestId())
                .isNotNull()
                .isEqualTo(persistedEntity.getId());
    }
}

File: database/src/test/kotlin/ca/metricalsky/winston/database/entity/fetch/FetchRequestEntityTest.kt
package ca.metricalsky.winston.database.entity.fetch

import ca.metricalsky.winston.database.entity.fetch.FetchRequestEntity.Status.ACCEPTED
import ca.metricalsky.winston.database.test.annotation.DatabaseTest
import ca.metricalsky.winston.database.test.faker.DatabaseFaker
import ca.metricalsky.winston.database.test.jpa.ext.refreshAll
import io.kotest.core.spec.style.WordSpec
import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager

@DatabaseTest
class FetchRequestEntityTest(
    entityManager: TestEntityManager
): WordSpec({

    val faker = DatabaseFaker();

    "FetchRequestEntity" should {

        "persist with only required fields" {
            val fetchRequest = faker.fetchRequest().minimalEntity()

            entityManager.persistFlushFind(fetchRequest) shouldNotBeNull {
                id shouldNotBe null
                status shouldBe ACCEPTED
                operations shouldBe emptyList()
                createdAt shouldNotBe null
                lastUpdatedAt shouldNotBe null
            }
        }

        "persist operations" {
            val operation = FetchOperationEntity()
            operation.operationType = FetchOperationEntity.Type.CHANNELS
            operation.objectId = faker.channel().id()

            val fetchRequest = FetchRequestEntity(
                operations = mutableListOf(operation)
            )

            entityManager.persistFlushFind(fetchRequest) shouldNotBeNull {
                id shouldNotBe null

                entityManager.refreshAll(operations)
                operations shouldHaveSize fetchRequest.operations.size
                operations.forEach {
                    it.id shouldNotBe null
                    it.fetchRequestId shouldBe id
                }
            }
        }
    }
})
